keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,287 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal is to remove the Container Runtime Interface (CRI) socket annotation  named `kubeadm.alpha.kubernetes.io/cri-socket` from the Node object in Kubernetes, which is currently added during the `init upload-config` phase in Kubeadm. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 
+
+instead of relying on the annotation, this KEP proses creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm init and join. This will eliminate the need for kubeadm to store CRI socket configuration on each Node object, instead relying on the kubeadm configuration file to specify this information.
+
+## Motivation
+
+Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.
+
+After migrating the container runtime endpoint flag to the instance configuration, we can use it
+Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.
+
+### Goals
+
+* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.
+* Provide an instance configuration file named `/var/lib/kubelet/instace-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.
+* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Non-Goals
+
+- Continue maintaining CRI socket paths on Node objects.
+
+## Proposal
+
+kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instance-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime endpoint will be automatically detected, uploaded to the global configuration, and `/var/lib/kubelet/instace-config.yaml` will be generated.
+
+kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
```

Comment:
```suggestion
kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration (`kube-system/kubelet-config`). If the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and `/var/lib/kubelet/instace-config.yaml` is generated based on it.
```

but here i'm confused, if the CRI socket is not specified in kubeadm config this happens, right?
- download the global config and put it in config.yaml
- autodetect the socket on this node
- write the detected socket to instance-config.yaml

the above comment in your text is not correct.

Reply:
Yes, this is a mistake, the steps you mentioned are more in line with our current process.